src/main/java/org/zalando/nakadi/service/FloodService.java

```diff
@@ -0,0 +1,177 @@
+package org.zalando.nakadi.service;
+
+import org.apache.commons.collections.map.HashedMap;
+import org.apache.curator.framework.recipes.cache.ChildData;
+import org.apache.curator.framework.recipes.cache.TreeCache;
+import org.apache.zookeeper.KeeperException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.zalando.nakadi.domain.EventType;
+import org.zalando.nakadi.exceptions.NakadiException;
+import org.zalando.nakadi.repository.db.EventTypeCache;
+import org.zalando.nakadi.repository.zookeeper.ZooKeeperHolder;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+@Component
+public class FloodService {
```

Comment:
> Do you think it is really harmful to use Curator's TreeCache for a big number of read operations? 

My observation is that extra load is being potentially being placed into a ZooKeeper that exist to run Kafka. It isn't just this feature, it's also being used iirc to store offsets for the hila (an approach modern Kafka has deliberately moved away from, by moving offsets into a Kafka topic). When running Kafka clusters at any kind of load of load or scale, the default operational approach is to give the Zk cluster exclusively to the Kafka system.

Put another way - if rate limiting was required by a service, would a ZooKeeper cluster be the automatic choice, or is it being used because it happens to be there and close to hand? 

Again - I do understand the urgency. I want to be sure a larger problem isn't being created for the future. So maybe this can ship, but there should be ongoing thinking whether operational stability is ultimately achieved in the mid/long term by placing multiple orthogonal workloads into ZooKeeper.